File: skills/xcrawl-search/scripts/search_query.js
#!/usr/bin/env node
const fs = require("node:fs");

function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith("--")) {
      args._.push(token);
      continue;
    }
    const eqIndex = token.indexOf("=");
    if (eqIndex !== -1) {
      const key = token.slice(2, eqIndex);
      args[key] = token.slice(eqIndex + 1);
      continue;
    }
    const key = token.slice(2);
    const next = argv[i + 1];
    if (next && !next.startsWith("--")) {
      args[key] = next;
      i += 1;
    } else {
      args[key] = "true";
    }
  }
  return args;
}

function buildPayload(args) {
  if (args["payload-file"] && args["payload-json"]) {
    throw new Error("Use only one of --payload-file or --payload-json");
  }
  if (args["payload-file"]) {
    return JSON.parse(fs.readFileSync(args["payload-file"], "utf8"));
  }
  if (args["payload-json"]) {
    return JSON.parse(args["payload-json"]);
  }

  return {
    query: args.query || args._[0] || "AI web crawler API",
    location: args.location || args._[1] || "US",
    language: args.language || args._[2] || "en",
    limit: Number(args.limit || args._[3] || 20),
  };
}

async function main() {
  const apiKey = process.env.XCRAWL_API_KEY;
  if (!apiKey) {
    console.error("XCRAWL_API_KEY is required");
    process.exit(1);
  }

  const args = parseArgs(process.argv.slice(2));
  const payload = buildPayload(args);

  const resp = await fetch("https://run.xcrawl.com/v1/search", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  const text = await resp.text();
  process.stdout.write(`${text}\n`);
  if (!resp.ok) {
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: skills/xcrawl-search/scripts/search_query.py
#!/usr/bin/env python3
import argparse
import json
import os
import sys

import requests


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Send a search request to Xcrawl")
    parser.add_argument("query", nargs="?", default=None, help="Query (positional shortcut)")
    parser.add_argument("--query", dest="query_opt", help="Query text")
    parser.add_argument("--location", default="US", help="Location code")
    parser.add_argument("--language", default="en", help="Language code")
    parser.add_argument("--limit", type=int, default=20, help="Result limit")
    parser.add_argument("--payload-file", help="Path to raw JSON payload file")
    parser.add_argument("--payload-json", help="Raw JSON payload string")
    return parser.parse_args()


def load_payload(args: argparse.Namespace) -> dict:
    if args.payload_file and args.payload_json:
        raise SystemExit("Use only one of --payload-file or --payload-json")

    if args.payload_file:
        with open(args.payload_file, "r", encoding="utf-8") as f:
            return json.load(f)

    if args.payload_json:
        return json.loads(args.payload_json)

    query = args.query_opt or args.query or "AI web crawler API"
    return {
        "query": query,
        "location": args.location,
        "language": args.language,
        "limit": args.limit,
    }


def main() -> None:
    api_key = os.getenv("XCRAWL_API_KEY")
    if not api_key:
        raise SystemExit("XCRAWL_API_KEY is required")

    payload = load_payload(parse_args())

    resp = requests.post(
        "https://run.xcrawl.com/v1/search",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        json=payload,
        timeout=60,
    )
    print(resp.text)
    if not resp.ok:
        sys.exit(1)


if __name__ == "__main__":
    main()
